peekaboo/sample.py

```diff
@@ -264,29 +267,59 @@ def known(self):
     @property
     def file_extension(self):
+        if self.has_attr('file_extension'):
+            return self.get_attr('file_extension')
+
+        # try to find a file name containing an extension. Using
+        # self.__filename will almost never yield anything useful because
+        # amavis intentionally hands us files named only p001, p002 and so on.
+        # But we still try it in case there's no declared name.
+        filename = self.__filename
         if self.has_attr('meta_info_name_declared'):
-            file_ext = self.get_attr('meta_info_name_declared').split('.')[-1]
-            if self.has_attr('file_extension'):
-                if self.get_attr('file_extension') != file_ext:
-                    self.set_attr('file_extension', file_ext, override=True)
-            else:
-                self.set_attr('file_extension', file_ext)
-        elif not self.has_attr('file_extension'):
-            file_ext = os.path.splitext(self.__filename)[1][1:]
-            self.set_attr('file_extension', file_ext)
-        return self.get_attr('file_extension')
+            filename = self.get_attr('meta_info_name_declared')
+
+        # extension or the empty string if none found
+        file_ext = os.path.splitext(filename)[1][1:]
+
+        # as a fallback try to guess file extension from mime type
+        if not file_ext:
+            for mt in self.mimetypes:
+                mtfe = guess_file_extension_from_mime_type(mt)
+                if mtfe:
+                    # strip leading dot
+                    file_ext = mtfe[1:]
+                    break
+
+        # file_ext will still be emtpy string if fallback couldn't guess any
+        # extension either
+        self.set_attr('file_extension', file_ext)
+        return file_ext
 
     @property
     def mimetypes(self):
-        """
-        Can not be cached (hard to determine if known/complete).
+        if self.has_attr('mimetypes'):
+            return self.get_attr('mimetypes')
 
-        determine mime on original p[0-9]* file
-        later result will be "inode/symlink"
-        """
-        mime_types = []
+        mime_types = set()
 
-        smime = {
+        # get MIME type from meta info
+        declared_mt = None
+        if self.has_attr('meta_info_type_declared'):
+            declared_mt = self.get_attr('meta_info_type_declared')
+            if declared_mt is not None:
+                logger.debug('Sample declared as "%s"' % declared_mt)
+                mime_types.add(declared_mt)
+
+        declared_filename = self.__filename
+        if self.has_attr('meta_info_name_declared'):
+            declared_filename = self.get_attr('meta_info_name_declared')
+
+        # check if the sample is an S/MIME signature (smime.p7s)
+        # If so, don't overwrite the MIME type since we do not want to analyse
+        # S/MIME signatures.
+        # FIXME: This is oddly specific for this generic routine. Should it be
+        # some sort of callback or plugin?
```

Comment:
This was put here to avoid false positives by scanning mail signaures (they are opened with ms certmgr, that triggers cuckoo signatures and hence the attachment is marked bad).  This has been put here rather than simply on the whitelist to combine the check with the filename.

I would accept this for now and include a better solution when refactoring the ruleset.

NOTE:
we lag support for pgp singatures here, they also cause false positives

Reply:
Agreed. Handled in #42.